src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,316 @@
+
+      ExecuteResponse response = null;
+      // Exit the loop as long as we get a response from either Execute() or WaitExecution().
+      while (response == null) {
+        // We use refreshIfUnauthenticated inside retry block. If use it outside, retrier will stop
+        // retrying when received a unauthenticated error, and propagate to refreshIfUnauthenticated
+        // which will then call retrier again. It will reset the retry time counter so we could
+        // retry more than --remote_retry times which is not expected.
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                executeBackoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implicated inside
+        // the retry block.
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              waitExecutionBackoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get().execute(request);
+
+        // We don't want to reset backoff for Execute() since if there is an error:
+        //   1. If happened before we received a first response, we want to ensure the retry counter
+        //      is increased and call Execute() again.
+        //   2. Otherwise, we will fallback to WaitExecution() loop.
+        return handleOperationStream(operationStream, (operation) -> {
+          lastOperation = operation;
+          observer.onNext(operation);
+        });
+      } catch (StatusRuntimeException e) {
+        if (lastOperation != null) {
+          // By returning null, we are going to fallback to WaitExecution() loop.
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(lastOperation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(lastOperation.getName())
+          .build();
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+
+        // We want to reset backoff for WaitExecution() so we can "infinitely" wait for the
+        // execution to complete as long as they are making progress (by returning response).
+        return handleOperationStream(operationStream, (operation) -> {
+          // Assuming the server has made progress since we received the response. Reset the
+          // backoff so that this request has a full deck of retries.
+          waitExecutionBackoff.reset();
+
+          lastOperation = operation;
```

Comment:
These two lines seem to be common between the two lambdas passed in the second arg of `handleOperationStream()`. Move them to where the lambda is called.

Reply:
Done.